---
title: "scRNAseq data analysis CRISPR Screening BM: MCA SingleR + Integration"
subtitle: 'Researcher: Luis Galan'
author: "Author: Eric Canton - Maria Maqueda"
date: "13th August 2024"
bibliography: "/home/user/Documents/Files/Projects/BM_scRNAseq/aux_files/references_scRNAseq_Analysis.bib"
output:
  html_document:
    df_print: paged
    css: /home/user/Documents/Files/Projects/BM_scRNAseq/aux_files/style.css
    toc: yes
    toc_depth: '3'
    toc_float: yes
    theme: cerulean
    highlight: haddock
  pdf_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
---

```{r packages, include=FALSE}
# Included any additional libraries required i.e. openxlsx
libraries <- c("knitr", "ggplot2", "data.table", "dplyr", "stringr", "ggpubr", "Seurat", "glmGamPoi", "clustree",       
               "multtest","metap","ComplexHeatmap", "RColorBrewer","ggpubr", "openxlsx", "readxl", 
               "ggvenn", "Rmagic", "monocle3", "SeuratWrappers", "patchwork", "gprofiler2") 
check.libraries <- is.element(libraries, installed.packages()[, 1])==FALSE
libraries.to.install <- libraries[check.libraries]
if (length(libraries.to.install!=0)) {
  require(libraries.to.install)
}

success <- sapply(libraries,require, quietly = FALSE,  character.only = TRUE)
if(length(success) != length(libraries)) {stop("A package failed to return a success in require() function.")}
```

```{r general, include=FALSE}
# General settings
knitr::opts_chunk$set(message=FALSE, warning=FALSE, echo=FALSE, fig.pos="H")
options(scipen=999)

# Define colors for conditions
cond.colors <- c("CD45_4_5" = "gray",
                 "CD45_BFP_4_5" = "gold3")

palette <- c("B cell" = "#A6CEE3", "Pre B cell" = "#1F78B4", 
             "Basophil" = "#B2DF8A", "Basophil progenitor" = "#33A02C",
             "Macrophage_Ctss high" = "#FB9A99", "Macrophage_Fcna high" = "#FB7CAD", "Macrophage_Ms4a6c high" = "#FB42A4", 
             "Myeloid cell" = "#DF1A1C",
             
             "Neutrophil_Chil3 high" = "#B2755A", "Neutrophil_Fcnb high" = "#FF8D3F", "Neutrophil_Ltf high" = "#FFC33F", 
             "Neutrophil_Mmp8 high" = "#EFE53E", "Neutrophil_Mpo high" = "#D5A86D", "Neutrophil progenitor" = "#D1BF24", 
             
             "T cell_Ccl5 high" = "#B070E8", "Innate lymphoid cell" = "#7070E8", 
             "Hematopoietic stem and progenitor cell" = "#6C7026", "Erythroid cell" = "#01F6AF", "Plasmacytoid dendritic cell" = "#613126",
             "Hematopoietic stem cell" = "#113226")

general.palette <- c("B cell" = "#A6CEE3",
             "Basophil" = "#B2DF8A",
             "Macrophage" = "#FB9A99",
             "Myeloid cell" = "#DF1A1C",
             "Neutrophil" = "#EFE53E",
             "T cell" = "#B070E8", 
             "ILC" = "#7070E8", 
             "HSPC" = "#113226", 
             "Erythroid cell" = "#01F6AF")
```

# Scope

This document aims to perform an integrated analysis on all available cells that passed the QC filtering. Results obtained in this analysis will determine if multilineage can be observed in any of the conditions. Samples have been previously been assigned a cell type using SingleR [@SingleR] and the (Bone Marrow) Mouse Cell Atlas [@MCA]. 

Sample integration is required since there is no perfect overlapping between both samples.

# Primary data

A total of **two samples** (Bone Marrow) were library-prepared and sequenced by CRG Sequencing platform at PRBB. Libraries were prepared using **10x Single Cell 3'v3 kit**. Corresponding biological samples were delivered to the sequencing platform in December'24 and sequencing data was received in February'24. Each sample corresponds to a different condition:

-   CD45+ cells at 4.5 months (**CD45_4_5**)
-   CD45+ cells transplanted from KDR+ progenitors at 4.5 months (**CD45_BFP_4_5**) activated from 7* specific genes identified by CRISPRa in an earlier project phase.

*Genes related to HSC development fate.

# Methods

Following criteria is applied:

-   SCTransform-based normalization (v2) regressing out mitochondrial content. Rest of parameters by default.
-   Dimensionality reduction considering 35 PCs which corresponds to a total cumulative variance explained of 85%.
-   Clustering (Louvain algorithm) with resolutions explored: from 0.1 to 0.6 (less to more specific).
-   'FindAllMarkers' was used to identify differentially expressed genes for each cluster and condition.

All previous steps and integration are performed by means of Seurat R package [@SingleR] (v5.0.0) together with glmGamPoi R package [@glmGamPoi] (v.1.12.2). Check vignette <https://satijalab.org/seurat/archive/v4.3/sctransform_v2_vignette>

# Results

```{r IMPORTANT}
# To execute this markdown quickly, load an RDS object already normalized + integrated + PCA + UMAP + Clustering + conserved cell markers
BM.combined.sct <-   readRDS(file="/home/user/Documents/Files/Projects/BM_scRNAseq/RDS/final/scRNAseq_CRISPR_Screening_BM_Annotated_Integrated_FINAL.RDS")
```


As a reminder, the dataset to be analyzed contains `r nrow(BM.combined.sct)` features (genes) over `r ncol(BM.combined.sct)` cells distributed in 2 samples from two conditions. In this case, ribosomal genes were excluded from the dataset. Low quality cells were also discarded as a result of a previous QC.

```{r, fig.width=7, fig.height=5}

Seurat::DimPlot(BM.combined.sct, 
                reduction = "umap", 
                group.by = "Sample_ID", 
                alpha=0.6,
                pt.size = 0.3,
                cols = cond.colors)

```

The UMAP shows a great overlap between samples after the integration.

```{r, fig.width=20, fig.height=11}

Seurat::DimPlot(BM.combined.sct, 
                reduction = "umap", 
                group.by = "cell.ident", 
                split.by = "Sample_ID",
                alpha=0.6,
                pt.size = 0.6,
                cols = palette)

```

This representation confirms that some cell types are predominant in the non transplanted cells (Macrophages) and others in the transplanted cells (B cells and T cells).

It is recommended to visualize UMAP against other possible confounding variables i.e. mitochondrial content or number of genes detected per cell. This is shown in next figures:

```{r, fig.width=16, fig.height=14}
# Color the dots according to percent.mito, ribo, nCount and nFeature
plots <- vector("list",length=4)
plots[[1]] <- Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = "percent.mito", pt.size = 0.15)
plots[[2]] <- Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = "percent.ribo", pt.size = 0.15)
plots[[3]] <- Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = "nCount_RNA", pt.size = 0.15)
plots[[4]] <- Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = "nFeature_RNA", pt.size = 0.15)
plt <- patchwork::wrap_plots(plots, ncol=2)
plt
```


## Cell phase

Finally, cell phase can be inferred from a reference set of genes defined by [@Tirosh] and phase labels represented in UMAP, as shown in following figure:

```{r}
mmus_s = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

seu <- Seurat::CellCycleScoring(BM.combined.sct,
                                s.features = mmus_s,
                                g2m.features = mmus_g2m)

Seurat::DimPlot(seu,
                     reduction = "umap",
                     group.by= "Phase")
```

Since this dataset includes developing cells, it is not recommended to regress out this information.

## Seurat Annotation

Different clustering resolutions are explored from lower (less specific cell type) to higher granularity (more specific cell type).

Clusters can be represented over UMAP, the following graph show this layout for clustering resolution 0.4.

```{r, fig.width=5, fig.height=5}
Seurat::DimPlot(BM.combined.sct, group.by = "integrated_snn_res.0.4" , label=TRUE)
```

**A clustering resolution of 0.4 is chosen**. Let's visualize the clusters per condition:

```{r, fig.width=16, fig.height=8}
plot <- vector("list",length=2)
plot[[1]] <- Seurat::DimPlot(BM.combined.sct, group.by = "integrated_snn_res.0.4" , label=TRUE)
plot[[2]] <- Seurat::DimPlot(BM.combined.sct, reduction =  "umap", group.by = "cell.ident", 
              cols = palette, alpha=0.5, pt.size = 0.25)
plt <- patchwork::wrap_plots(plot, ncol=2)
plt
```

The distribution of cells among clusters and between group is:

```{r config , fig.width=12, fig.height=12}
print(table(BM.combined.sct@meta.data$integrated_snn_res.0.4, 
            BM.combined.sct$Sample_ID))

```


## Subclustering

There are some cell types not present in the Mouse Cell Atlas that can be observed using different Hematopoietic markers.
The manual re annotation process is as follows:
-   Plot markers of an specific cell type and observe positive cells in one seurat cluster (res 0.4) 
-   If all the cluster is positive for this marker, it is re annotated.
-   If the markers are partially expressed, a subcluster is calculated to fit the positive cells.

## MAGIC imputation

MAGIC [@MAGIC] is an imputation algorithm used to restore the structure of the data. We will use this tool to visualize the expression of different hematopoietic markers. This is an example of how the algorithm works:

The plot on the right shows how the imputation has revealed the expression of the gene "Gata3" in one cluster.

```{r, fig.width=15, fig.height=7}
plot <- vector("list",length=2)
DefaultAssay(BM.combined.sct) <- 'SCT'
plot[[1]] <- FeaturePlot(BM.combined.sct, reduction = "umap", features = "Gata3", pt.size = 0.15) & scale_color_viridis()
DefaultAssay(BM.combined.sct) <- 'MAGIC_SCT'
plot[[2]] <- FeaturePlot(BM.combined.sct, reduction = "umap", features = "Gata3", pt.size = 0.15) & scale_color_viridis()
plt <- patchwork::wrap_plots(plot, ncol=2, )
plt

```

## Hematopoietic markers

From Kucinski et al, 2024: <https://doi.org/10.1016/j.stem.2023.12.001>

```{r}
DefaultAssay(BM.combined.sct) <- 'MAGIC_SCT'
```

### HSC_markers

```{r HSC_markers UMAP, fig.width=15, fig.height=12}
HSC_m =c("Hlf","Mecom", "Procr")
Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = HSC_m, pt.size = 0.1) & scale_color_viridis()

```

This signature is expressed in cluster 11, and therefore it has been tagged as *"Hematopoietic stem cell"*.

### Neutrophil_Progenitor_markers

```{r Neutrophil_Progenitor_markers UMAP, fig.width=15, fig.height=12}
Neutrophil_Progenitor_m =c("Elane", "Prtn3", "Ms4a3")
Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = Neutrophil_Progenitor_m, pt.size = 0.1) & scale_color_viridis()

```

This signature is expressed in cluster 9, and therefore it has been tagged as *"Neutrophil progenitor"*.

### Basophil_Progenitor_markers

```{r Basophil_Progenitor_markers UMAP, fig.width=15, fig.height=7}
Basophil_Progenitor_m= c("Prss34", "Mcpt8")
Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = Basophil_Progenitor_m, pt.size = 0.1) & scale_color_viridis()

```

This signature is expressed partially in cluster 11. After subclustering, cluster 11_6 fits the positive cells, and therefore it has been tagged as *"Basophil progenitor"*.

### ILC_markers

```{r ILC_m UMAP, fig.width=15, fig.height=12}
ILC_m=c("Gata3", "Id2", "Il7r", "Irf3")
Seurat::FeaturePlot(BM.combined.sct, reduction = "umap", features = ILC_m, pt.size = 0.1) & scale_color_viridis()
```

This signature is consistently expressed in cluster 14, and therefore it has been tagged as *"Innate lymphoid cell"*.


### Curated.cell.ident

After re annotation, cell types are represented as follows:

```{r, fig.width=18, fig.height=11}
plot <- vector("list",length=2)
plot[[1]] <- Seurat::DimPlot(BM.combined.sct, 
               reduction =  "umap", 
               group.by = "curated.cell.ident", alpha=0.5,
               cols = palette,
               pt.size = 0.6) + NoLegend()

seu <- Seurat::SetIdent(BM.combined.sct, value = BM.combined.sct$curated.cell.ident)
pt <- table(Idents(seu), BM.combined.sct$Sample_ID)
pt <- as.data.frame(pt)
pt$Var1 <- as.character(pt$Var1)

plot[[2]] <- ggplot(pt, aes(x = Var2, y = Freq, fill = Var1)) +
  theme_bw(base_size = 15) +
  geom_col(position = "fill", width = 0.5) +
  xlab(" ") +
  ylab("Proportion") +
  scale_fill_manual(values = palette) +
  theme(legend.title = element_blank(),
    legend.text = element_text(size=10, face = "bold"),
        axis.text.x=element_text(size=15, face = "bold"),
        axis.text.y=element_text(size=15, face = "bold"),
        axis.title = element_text(size=20,face = "bold")) 

plt <- patchwork::wrap_plots(plot, ncol=2, )
plt


```

```{r fig.width=18, fig.height=10}
Seurat::DimPlot(BM.combined.sct, 
                reduction = "umap", 
                group.by = "curated.cell.ident", 
                split.by = "Sample_ID",
                alpha=0.6,
                pt.size = 0.6,
                cols = palette)

```


## General Cell Type Annotation

These 20 cell types were summarized into 9 main cell types (B cell, Basophil, Erythroid, HSC/HSPC, ILC, Macrophage, Myeloid cells, Neutrophil and T cells).

```{r, fig.width=18, fig.height=11}
plot <- vector("list",length=2)
plot[[1]] <- Seurat::DimPlot(BM.combined.sct, 
               reduction =  "umap", 
               group.by = "general.curated.cell.ident", alpha=0.5,
               cols = general.palette,
               pt.size = 0.65) + NoLegend()

seu <- Seurat::SetIdent(BM.combined.sct, value = BM.combined.sct$general.curated.cell.ident)
pt <- table(Idents(seu), BM.combined.sct$Sample_ID)
pt <- as.data.frame(pt)
pt$Var1 <- as.character(pt$Var1)

plot[[2]] <- ggplot(pt, aes(x = Var2, y = Freq, fill = Var1)) +
  theme_bw(base_size = 15) +
  geom_col(position = "fill", width = 0.5) +
  xlab(" ") +
  ylab("Proportion") +
  scale_fill_manual(values = general.palette) +
  theme(legend.title = element_blank(),
    legend.text = element_text(size=10, face = "bold"),
        axis.text.x=element_text(size=15, face = "bold"),
        axis.text.y=element_text(size=15, face = "bold"),
        axis.title = element_text(size=20,face = "bold")) 

plt <- patchwork::wrap_plots(plot, ncol=2, )
plt


```


## Seurat Re-Annotation

Finally, clusters 11, 12, 13 and 15 were subclustered to fit the final cell types.

```{r}

Seurat::DimPlot(BM.combined.sct, group.by = "subcluster.integrated_snn_res.0.4", label=TRUE)

```

```{r}

Seurat::DimPlot(BM.combined.sct, group.by = "final.clusters.integrated_snn", label=TRUE)

```

# References

