
################################################################################
##        Title : scRNAseq data analysis (EBs 7g)
##  Description : This script is for conducting differential expression analysis 
##                from scRNAseq - pseudobulk approach
##   Researcher : Luis Galán Palma
##       Author : María Maqueda
##         Date : 31st July 2024
################################################################################

## NOTE: This pseudobulk analysis is applied over Kdr+ cells. Based on the complete
## dataset, those cells showing at least 1 raw count are subsetted. Check .Rmd
## script (final section) where the subsetting is conducted.

# Basically, this script aggregates the expression of selected cells per sample
# and then performs and standard DEA with DESeq2 (as the one applied for bulk)

################################################################################
## 1. Load packages
################################################################################

library(Seurat) # To import original RDS object with Seurat object
library(openxlsx)

require(dplyr)
require(reshape2)
require(DESeq2) # For diff analysis
require(apeglm) # To shrink logFCs
require(vsn) # To plot mean-sd plots
require(EnhancedVolcano)
require(ggpubr) # For arranging several plots in one
require(UpSetR)
require(ggplot2)
require(ggrepel)

# Custom script for retrieving results. DO NOT EDIT THIS FILE, just uploaded into
source("/Volumes/grcmc/ABigas_Bioinfo/SCRIPTS/RNA-seq/source/Misc_scripts_RNAseq_analysis_v0.R")


################################################################################
## 2. Import data: scRNAseq dataset to be tested and prepare raw count matrix
################################################################################

# scRNAseq dataset Kdr positive: Check Rmd (clustering) where this RDS is created
seu <- readRDS("RDS/scRNAseq_EBs_SUBSET_KDRpos_FINAL.RDS")

# Create metadata information with sample name (orig.ident) and corresponding Group 
metadata <- seu@meta.data[,c("orig.ident", "Group")]

metadata <- dplyr::distinct(metadata) # Otherwise, we have all list of cells
rownames(metadata) <- metadata$orig.ident # Metadata rownames must match sample names in count matrix

metadata
# orig.ident Group
# WT_s1      WT_s1    WT
# WT_s2      WT_s2    WT
# WT_s3      WT_s3    WT
# g7_s1      g7_s1    g7
# g7_s2      g7_s2    g7
# g7_s3      g7_s3    g7

metadata$Group <- relevel(as.factor(metadata$Group),ref="WT") #To ensure test g7 vs WT IMPORTANT!!

raw.count.matrix <- as.data.frame(Seurat::AggregateExpression(object = seu, 
                                  assays = "RNA", # We want raw counts
                                  group.by = "orig.ident",
                                  return.seurat =FALSE))
# colSums(raw.count.matrix)

colnames(raw.count.matrix) <- gsub(pattern = "RNA.", replacement = "", x= colnames(raw.count.matrix))
colnames(raw.count.matrix) <- gsub(pattern = "\\.", replacement = "_", x= colnames(raw.count.matrix))

# > colnames(raw.count.matrix)
# [1] "g7_s1" "g7_s2" "g7_s3" "WT_s1" "WT_s2" "WT_s3"

# Now sample names match!!
# We can put the metadata in the same order

metadata <- metadata[match(metadata$orig.ident, colnames(raw.count.matrix)),]

################################################################################
## 4. Create DESeqDataSet object 
################################################################################

# Generate the DESeqDataSet from pseudobulk matrix (keep)
# https://github.com/tavareshugo/tutorial_DESeq2_contrasts/blob/main/DESeq2_contrasts.md
data.dds <- DESeq2::DESeqDataSetFromMatrix(countData = raw.count.matrix, #counts.subset
                                            colData = metadata, #metadata.subset
                                            design =  ~ Group,
                                            tidy = FALSE)  # We've already add the rownames when importing data
 
data.dds
dim(data.dds) # 22492 genes x 6 samples [this dimension will be the same for all cases]
 
 # Remove low-count genes: those with less than 5 counts across all samples (not needed, all genes have > 5 counts) 
data.dds.red <- data.dds[rowSums(counts(data.dds)) >= 5] 
dim(data.dds.red) 

################################################################################
## 5. Mean-Sd plots to inspect normalization effect
################################################################################
 
# Raw counts: As expected, high standard dev at higher count ranges
g1 <- meanSdPlot(assay(data.dds.red))
g1 <- g1$gg + ggtitle("RAW counts")
 
# We choose to use blind=FALSE since we do not expect the majority of regions to have
# large counts differences between conditions
# VST
vsd <- vst(data.dds.red, blind=FALSE)
g2 <- meanSdPlot(assay(vsd))
g2 <- g2$gg + ggtitle("VST counts")
 
# RLOG
rld <- rlog(data.dds.red, blind=FALSE)
g3 <- meanSdPlot(assay(rld))
g3 <- g3$gg + ggtitle("RLOG counts")
 
 # CONCLUSION: VST is the best performing
 
pdf(file = "PAPER_RELATED/Pseudobulk_DEA/Kdr_positive/MeanSD_plots_diffNorm_methods_Kdr_pos.pdf", height=4, width=10)
ggarrange(g1, g2, g3, 
           labels = c("A", "B", "C"),
           ncol = 3, nrow = 1)
dev.off()
 
 
 ## Best normalization : VST
 

################################################################################
## 7. Plot: PCA (VST normalization)
################################################################################
 
# Function to plot PCAs 
PCA <- PCA_custom(Norm_Matrix = vsd,
                   TopGenes = NULL,          #If TopGenes = NULL, all genes
                   filename = "Pseudobulk_PCA_all_genes_Kdr_pos",
                   title = "PCA All genes",
                   ComponentX = "PC1",
                   ComponentY = "PC2",
                   colorby = "Group",        #Leave NULL not to classify by color
                   shapeby = NULL,    #Leave NULL not to classify by shape
                   colors = c("WT" = "lightblue", "g7" = "indianred"),
                   xlim = NULL,
                   ylim = NULL,
                   res_path = "PAPER_RELATED/Pseudobulk_DEA/Kdr_positive/")

 

################################################################################
## 8. Boxplot of expression data
################################################################################
 
toplot <- melt(assay(vsd))
colnames(toplot) <- c("Gene","Sample","Transformed_Counts")
 
toplot <- toplot %>%
   mutate(Group = case_when(Sample == "g7_s1" | Sample == "g7_s2" | Sample == "g7_s3"  ~ "g7",
                            TRUE ~ "WT"))  
 
pdf(file = "PAPER_RELATED/Pseudobulk_DEA/Kdr_positive/Boxplot_all_genes_Kdr_positive.pdf",width=10, height=7)
ggplot(toplot, aes(x=Sample, y=Transformed_Counts, fill=Group)) + 
   geom_violin() +
   theme_bw() +
   facet_wrap(~Group, scales="free_x") +
   scale_fill_manual(breaks = c("WT", "g7"),
                     values=c("lightblue", "indianred")) +
   stat_summary(fun=median, geom="point", size=2, color="grey")
dev.off()
 
################################################################################
## 10. Conduct Diff Expr Analysis and retrieve results
################################################################################

res_path="PAPER_RELATED/Pseudobulk_DEA/Kdr_positive" # Function parameter not working properly
# (NOTE: a /figures and /tables subfolder must be present in this path)
 
# Conduct analysis
dea <- DESeq(data.dds.red)

resultsNames(dea)  #Identify the name of the comparison of interest
# [1] "Intercept"      "Group_g7_vs_WT"

tests_of_interest <- list("g7_vs_WT" = "Group_g7_vs_WT")

# Retrieve results 
final.results <- lapply(seq(1,length(tests_of_interest)), function(case){
  
  retrieve_final_results(DESeqDataSet_res = dea, 
                         contrast.or.coef = "coef",
                         test = tests_of_interest[[case]],
                         comparison.name = "g7_vs_WT_Kdr_pos_neigh_sign",
                         lFC_shrinkage=TRUE,
                         alpha = 0.05,
                         lFC = 0,
                         res_path = "PAPER_RELATED/Pseudobulk_DEA/Kdr_positive",
                         gene_annotations = NULL
  )})

