---
title: "scRNAseq data analysis (EBs 7g): Mapping summary from CellRanger"
subtitle: 'Researcher: Luis García Palma'
author: "Author: María Maqueda"
date: "13th May 2024"
bibliography: "/Users/mmaqueda/Documents/Projects/Luis_Galán/CRISPR_Screening/scRNAseq_EBs/aux_files/references_scRNAseq_Analysis.bib"
output:
  html_document:
    df_print: paged
    css: style.css
    toc: yes
    toc_depth: '3'
    toc_float: yes
    theme: cerulean
    highlight: haddock
  pdf_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
---

```{r packages, include=FALSE}
# Included any additional libraries required i.e. openxlsx
libraries <- c("knitr", "ggplot2", "data.table", "dplyr", "stringr", "ggpubr") 
check.libraries <- is.element(libraries, installed.packages()[, 1])==FALSE
libraries.to.install <- libraries[check.libraries]
if (length(libraries.to.install!=0)) {
  require(libraries.to.install)
}

success <- sapply(libraries,require, quietly = FALSE,  character.only = TRUE)
if(length(success) != length(libraries)) {stop("A package failed to return a success in require() function.")}
```

```{r general, include=FALSE}
# General settings
knitr::opts_chunk$set(message=FALSE, warning=FALSE, echo=FALSE, fig.pos="H")
options(scipen=999)

# Project path in IMIM cluster where CellRanger outputs is located
cluster.path <- "/Volumes/cancer/CRISPR_screening_mESCs_ABigas/scRNAseq_EBs_7g_vs_WT/cellranger_out/"

# Define colors for conditions: WT and 7g
cond.colors <- c("WT" = "lightblue",
                  "g7" = "indianred")
```


# Scope

This document is to check the quality of the mapping performed with CellRanger tool from 10x Genomics. Specifically, visualization of the common mapping metrics obtained per library (sample) are considered.

# Primary data

A total of **six samples** were library-prepared and sequenced by NovoGene Co, Ltd. Libraries were prepared using **10x Single Cell 3'v3 kit**. Corresponding biological samples were delivered to IJC at the end of October'23 and sequencing data was received at the end of February'24. Samples are equally divided in two conditions, considering cells from **mouse embryoid bodies (mEBs E6.5)** :

-   in wild-type (**WT**) condition (3x) and,
-   after activation of 7* specific genes (**7g**) (3x) identified by CRISPRa in an earlier project phase.

*Genes related to HSC development fate.

# Methods

Visualized results in this report have been generated by CellRanger v7.2.0 [@CellRanger] in a HPC environment using Singularity v3.8.3. Specifically, `cellranger count` was executed per sample to generate the corresponding data matrices. **Introns were included in expression quantification** as recommended by 10x Genomics (See for additional information https://www.10xgenomics.com/support/software/cell-ranger/latest/miscellaneous/cr-intron-mode-rec). Samples were aligned to *refdatagex-
mm10-2020-A* transcriptome provided as well by 10x Genomics. 

# Results

## Barcode Rank Plot

All six samples were successfully mapped to reference. Following figure shows the **Barcode rank plot** per sample. This plot shows the distribution of UMI counts in barcodes plotted in decreasing order of their corresponding number of UMIs. **All samples show a typical (and good) behaviour** with a 'cliff and knee' shape.  This is indicative of good separation between cell-associated barcodes and barcodes with empty emulsions. For more information on this, check: https://cdn.10xgenomics.com/image/upload/v1660261286/support-documents/CG000329_TechnicalNote_InterpretingCellRangerWebSummaryFiles_RevA.pdf


![](./aux_files/CR_Barcode_Rank_Plots.png){width="700px"}

## Overall statistics

Following metrics, per sample, are visualized:

- general values: estimated number of cells, mean reads per cell and median genes per cell.
- related to mapping quality: reads confidently (uniquely) map to intergenic/intronic or exonic regions.
- sequencing saturation

```{r import stats data}
# Read all csv files (one per sample)
samples <- c("WT_s1", "WT_s2", "WT_s3", "g7_s1", "g7_s2", "g7_s3")

data <- lapply(samples, function(s){
  a <- read.delim(file = paste0(cluster.path, s, "/outs/metrics_summary.csv"),
                header = TRUE,dec = ".", sep=",")
  a$Sample <- s # Add the name of the sample
  return(a[,c(20,1:19)])
})
data <- do.call(rbind, data)

data$Condition <- "WT"
data$Condition[grep("g7", data$Sample)] <- "g7"

```


### General values: cells, reads and genes

The **estimated number of cells** per sample is:

```{r number_cells}

data$Estimated.Number.of.Cells <- as.integer(gsub(pattern = ",", replacement = "", x=data$Estimated.Number.of.Cells))

ggplot(data, aes(x=as.factor(Sample), y=Estimated.Number.of.Cells, fill=Condition)) +
  geom_bar(stat="identity",width=0.8) +
  scale_fill_manual(values =  cond.colors) +
  xlab(" ") + ylab("Estimated Number Cells") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border=element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(),
        axis.text.y = element_text(size=12, face="bold"),
        axis.title.y =  element_text(size=12, face = "bold"),
        axis.text.x = element_text(size=12,face = "bold" ,angle = 45, vjust=0.5)) 

```

The average number of cells for the **WT** samples is `r round(mean(data$Estimated.Number.of.Cells[grep("WT", data$Sample)]), digits=0)` cells, while for the **g7** group is `r round(mean(data$Estimated.Number.of.Cells[grep("g7", data$Sample)]), digits=0)` cells. Thus, the **number of cells in both conditions is very similar**.


The **median genes per cell** per sample is:

```{r genes_per_cell}

data$Median.Genes.per.Cell <- as.integer(gsub(pattern = ",", replacement = "", x=data$Median.Genes.per.Cell))

ggplot(data, aes(x=as.factor(Sample), y=Median.Genes.per.Cell, fill=Condition)) +
  geom_bar(stat="identity",width=0.8) +
  scale_fill_manual(values =  cond.colors) +
  xlab(" ") + ylab("Median Genes per Cell") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border=element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(),
        axis.text.y = element_text(size=12, face="bold"),
        axis.title.y =  element_text(size=12, face = "bold"),
        axis.text.x = element_text(size=12,face = "bold" ,angle = 45, vjust=0.5)) 

```
The average number of median genes per cell for the **WT** samples is `r round(mean(data$Median.Genes.per.Cell[grep("WT", data$Sample)]), digits=0)` genes, while for the **g7** group is `r round(mean(data$Median.Genes.per.Cell[grep("g7", data$Sample)]), digits=0)` genes. Thus, the **median number of genes in both conditions is very similar**.


The **mean reads per cell** per sample is:

```{r reads_per_cell}

data$Mean.Reads.per.Cell <- as.integer(gsub(pattern = ",", replacement = "", x=data$Mean.Reads.per.Cell))

ggplot(data, aes(x=as.factor(Sample), y=Mean.Reads.per.Cell, fill=Condition)) +
  geom_bar(stat="identity",width=0.8) +
  scale_fill_manual(values =  cond.colors) +
  xlab(" ") + ylab("Mean Reads per Cell") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border=element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(),
        axis.text.y = element_text(size=12, face="bold"),
        axis.title.y =  element_text(size=12, face = "bold"),
        axis.text.x = element_text(size=12,face = "bold" ,angle = 45, vjust=0.5)) 

```

The average number of median genes per cell for the **WT** samples is `r round(mean(data$Mean.Reads.per.Cell[grep("WT", data$Sample)]), digits=0)` genes, while for the **g7** group is `r round(mean(data$Mean.Reads.per.Cell[grep("g7", data$Sample)]), digits=0)` genes. Thus, the **mean reads per cell in both conditions is very similar**.


### Mapping quality: Reads confidently map to intergenic, intronic or exonic regions

In this section, percentage of reads mapped to intergenic, intronic or exonic regions of the reference are shown. Reads mapping to intergenic regions may be associated to contamination of ambient RNA. In this analysis, this potential contamination is not quantified but, in general terms, a 5% of contamination rate is usual. Mapping to intronic regions may be associated to unspliced RNA or mature RNA with retained introns.


```{r mapping_quality}
# Reads mapped to genome and transcriptome
mapping_qc_vars <- c(
  "Reads.Mapped.Confidently.to.Genome",
  "Reads.Mapped.Confidently.to.Intergenic.Regions",
  "Reads.Mapped.Confidently.to.Intronic.Regions",
  "Reads.Mapped.Confidently.to.Exonic.Regions")

# Remove their %
for (col in colnames(data)) {
  if (any(str_detect(data[, col], "%"))) {
    data[, col] <- as.double(str_remove(data[, col], "%"))
  }
}

mapping_qc <- purrr::map(mapping_qc_vars, function(var) {
  ggplot(data, aes_string(x="Sample", y=var, fill = "Condition")) +
    geom_bar(stat="identity",width=0.8) +
    scale_y_continuous(limits = c(0, 100)) +
    scale_fill_manual(values =  cond.colors) +
    xlab(" ") + 
    ylab(paste0(unlist(str_split(string = var, pattern = "Reads.Mapped.Confidently.to."))[2], " %")) +
    theme_minimal() +
    theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border=element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(),
        axis.text.y = element_text(size=12, face="bold"),
        axis.title.y =  element_text(size=12, face = "bold"),
        axis.text.x = element_text(size=12,face = "bold" ,angle = 45, vjust=0.5)) 
})

# Plot all figures
ggarrange(plotlist = mapping_qc, nrow = 2, ncol = 2)


```

Previous graphs show **very good mapping metrics for all samples**, in fact, all of them behave quite similarly. Mapped reads to intergenic regions is contained (around 5%) and  to genome is high (>86%). In general, for mouse we would expect >85% mapped reads to genome for a good experiment.


### Sequencing saturation

Sequencing saturation is a measure of the fraction of library complexity that was captured during sequencing.

```{r seq saturation}

ggplot(data, aes(x=as.factor(Sample), y=Sequencing.Saturation, fill=Condition)) +
  geom_bar(stat="identity",width=0.8) +
  scale_fill_manual(values =  cond.colors) +
  xlab(" ") + ylab("Seq Saturation %") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border=element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(),
        axis.text.y = element_text(size=12, face="bold"),
        axis.title.y =  element_text(size=12, face = "bold"),
        axis.text.x = element_text(size=12,face = "bold" ,angle = 45, vjust=0.5)) 

```

For this experiment, sequencing saturation ranges between 21% and 25%. Required percentages will depend on the objective of the experiment *i.e.* these values would not be sufficient to detect low/very low expressed transcripts.


# Session Information

```{r}
sessionInfo()
```

# References