---
title: "scRNAseq data analysis (EBs 7g): Filtering low quality cells and genes - Final Round QC"
subtitle: 'Researcher: Luis García Palma'
author: "Author: María Maqueda"
date: "16th May 2024"
bibliography: "/Users/mmaqueda/Documents/Projects/Luis_Galán/CRISPR_Screening/scRNAseq_EBs/aux_files/references_scRNAseq_Analysis.bib"
output:
  html_document:
    df_print: paged
    css: style.css
    toc: yes
    toc_depth: '3'
    toc_float: yes
    theme: cerulean
    highlight: haddock
  pdf_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
---

```{r packages, include=FALSE}
# Included any additional libraries required i.e. openxlsx
libraries <- c("knitr", "ggplot2", "data.table", "dplyr", "stringr", "ggpubr", "Seurat") 
check.libraries <- is.element(libraries, installed.packages()[, 1])==FALSE
libraries.to.install <- libraries[check.libraries]
if (length(libraries.to.install!=0)) {
  require(libraries.to.install)
}

success <- sapply(libraries,require, quietly = FALSE,  character.only = TRUE)
if(length(success) != length(libraries)) {stop("A package failed to return a success in require() function.")}
```

```{r general, include=FALSE}
# General settings
knitr::opts_chunk$set(message=FALSE, warning=FALSE, echo=FALSE, fig.pos="H")
options(scipen=999)

# Define colors for conditions: WT and 7g
cond.colors <- c("WT" = "lightblue",
                  "g7" = "indianred")

# Define sample names
samples <- c("WT_s1", "WT_s2", "WT_s3", "g7_s1", "g7_s2", "g7_s3")
```

# Scope

This document aims to filter out poor quality cells and genes. This is an standard first QC step for the scRNAseq data pre-processing analysis. Commonly used QC metrics include the exploration of number of unique genes detected in each cell or the percentage of reads that map to the mitochondrial genome.

# Primary data

A total of **six samples** were library-prepared and sequenced by NovoGene Co, Ltd. Libraries were prepared using **10x Single Cell 3'v3 kit**. Corresponding biological samples were delivered to IJC at the end of October'23 and sequencing data was received at the end of February'24. Samples are equally divided in two conditions, considering cells from **mouse embryoid bodies (mEBs E6.5)** :

-   in wild-type (**WT**) condition (3x) and,
-   after activation of 7* specific genes (**7g**) (3x) identified by CRISPRa in an earlier project phase.

*Genes related to HSC development fate.

# Methods

Following criteria is applied:

- Cells with mitochondrial content higher than **10%** are discarded.
- Cells with less than 1k counts are discarded.
- Cells with more than 7k genes or less than 300 genes are discarded.

Ribosomal genes are removed from the dataset. Additionally, those genes expressed in 10 or less cells (among all samples) are also discarded.

QC is performed by means of Seurat R package [@SeuratV5] (v5.0.0).

REMARK: Cells classified as doublets/multiplets per sample, identified in a previous analysis, will be also discarded in this analysis.

```{r}
# QC criteria
min_nCount <- 1000
min_n_genes <- 300
max_n_genes <- 7000
max_mito <- 10
min_cells <- 10
```


# Results

All scRNAseq data samples were merged into a Seurat object

```{r import Seurat object}
raw.data <- readRDS(file = "RDS/All_samples_original_data.RDS")
raw.data
```

The complete dataset, without any filtering applied, includes a total of **32,285 features (genes)** over **47,177 samples (cells)** distributed in 6 samples from two conditions.

```{r cells distribution}
qc.summary <- as.data.frame(table(raw.data$Sample_ID))
colnames(qc.summary) <- c("Sample.ID", "Initial.Cells")
```


## Doublets removal 

Prior to explore QC quality level of the cells, those already classified as doublets are removed. Doublets identification was independently conducted per sample.

```{r doublets removal}
raw.data.woDbl <- subset(x =  raw.data, subset = scDblFinder.class == "singlet")
qc.summary$Cells.woDBL <- table(raw.data.woDbl$Sample_ID)

knitr::kable(qc.summary, caption="Number of cells per sample: initially and after doublets removal")
```

## QC metrics

```{r include QC proportions in metadata}

# MITOCHONDRIAL genes
raw.data.woDbl <- Seurat::PercentageFeatureSet(raw.data.woDbl, 
                                    pattern = "^mt-", 
                                    col.name = "percent.mito")

# RIBOSOMAL genes
raw.data.woDbl <- Seurat::PercentageFeatureSet(raw.data.woDbl, 
                                    pattern = "^Rp[sl]",
                                    col.name = "percent.ribo")

```


### Exploration 

Prior to apply any QC filtering, first a general exploration is conducted to assess the overall cells quality. This includes, per barcode (cell):

- Number of genes
- Number of counts
- Mitochondrial content 
- Ribosomal content

```{r QC exploration 1}
Seurat::VlnPlot(raw.data.woDbl, layer = "counts",
        features = c("nFeature_RNA", "nCount_RNA"), 
        alpha=0.1, 
        ncol = 2, cols=c(rep("lightblue",3), rep("indianred",3)))
```

```{r QC exploration 2}
Seurat::VlnPlot(raw.data.woDbl, layer = "counts",
        features = c("percent.mito", "percent.ribo"), 
        alpha=0.1, 
        ncol = 2, cols=c(rep("lightblue",3), rep("indianred",3)))
```

Alternatively, a joint plot between metrics is also of interest to check relationships among variables. 

```{r joint QC}
ggplot(raw.data.woDbl@meta.data, 
       aes(x=nCount_RNA, y=nFeature_RNA , 
           color=percent.mito)) +
  geom_point() +
  scale_x_log10(labels = scales::comma) +
  scale_y_log10(labels = scales::comma) +
  theme_bw() +
  ggtitle("Features detected vs Library Size")  +
  facet_wrap(~orig.ident) 
```

All samples behave roughly equally independently of their condition.

### Removal criteria with MADs

Explore possible cutoffs based on four MADs:

```{r, eval=FALSE}
### Compute thresholds for minimum allowed values for number of UMIs and Genes

# 1) Include the log1p values for nCount and nFeature in the metadata
raw.data.woDbl$log1p_nCount_RNA <- log1p(raw.data.woDbl$nCount_RNA)
raw.data.woDbl$log1p_nFeature_RNA <- log1p(raw.data.woDbl$nFeature_RNA)

# Function to define lower and upper limits
define_limits <- function(metric,nmads,returnlog1p=FALSE){
  lower <- median(metric) - nmads*mad(metric)
  upper <- median(metric) + nmads*mad(metric)
  if(returnlog1p){
    return(list(lower, upper))
  }else{
    return(list(exp(lower)+1, exp(upper)+1))
  }}

# Ncount Lower limit: a minimum of 1000 is reasonable
sapply(samples, function(s) {
  define_limits(metric= raw.data.woDbl$log1p_nCount_RNA[which(raw.data.woDbl$orig.ident %in% s)], nmads=4)[[1]]
} )

#     WT_s1     WT_s2     WT_s3     g7_s1     g7_s2     g7_s3 
# 1386.8945 1439.4392 1002.3997  799.4377 1167.7756 1089.5985

# Nfeature Lower Limit: Based on this, 750 genes would enclose all cases. However, we will apply 300 as criteria and rectified
# if necessary in followin analysis steps
sapply(samples, function(s) {
  define_limits(metric= raw.data.woDbl$log1p_nFeature_RNA[which(raw.data.woDbl$orig.ident %in% s)], nmads=4)[[1]]
} )

#     WT_s1     WT_s2     WT_s3     g7_s1     g7_s2     g7_s3 
# 1162.8045 1181.0766  874.4644  758.7555  985.4532 1023.4013 


# Mito content Higher Limit: To be more conservative and include all cases: 10% for all
sapply(samples, function(s) {
  define_limits(metric= raw.data.woDbl$percent.mito[which(raw.data.woDbl$orig.ident %in% s)], nmads=4, returnlog1p = TRUE)[[2]]
} )

#    WT_s1    WT_s2    WT_s3    g7_s1    g7_s2    g7_s3 
# 11.57391 10.77505 10.38777 11.39322 11.59207 12.35611 

```


### Subset low quality cells

Cells are subset based on abovementioned QC criteria. The final number of cells is shown in the following table.

```{r subset low quality cells}
is_low_quality <- 
  raw.data.woDbl$nCount_RNA < min_nCount |
  raw.data.woDbl$nFeature_RNA < min_n_genes |
  raw.data.woDbl$nFeature_RNA > max_n_genes |
  raw.data.woDbl$percent.mito > max_mito 

#table(is_low_quality)
raw.data.woDbl.qcFilt <- subset(raw.data.woDbl, cells = colnames(raw.data.woDbl)[!is_low_quality])

qc.summary$Cells.FINAL <- table(raw.data.woDbl.qcFilt$Sample_ID)
knitr::kable(qc.summary, caption="Number of cells per sample: initially, after doublets removal and after final QC criteria")
```

Of note, most of the cells discarded is because of the mitochondrial percentage. After filtering, QC plots are again visualized:

```{r QC exploration 1 after}
Seurat::VlnPlot(raw.data.woDbl.qcFilt, layer = "counts",
        features = c("nFeature_RNA", "nCount_RNA"), 
        alpha=0.1, 
        ncol = 2, cols=c(rep("lightblue",3), rep("indianred",3)))
```

```{r QC exploration 2 after}
Seurat::VlnPlot(raw.data.woDbl.qcFilt, layer = "counts",
        features = c("percent.mito", "percent.ribo"), 
        alpha=0.1, 
        ncol = 2, cols=c(rep("lightblue",3), rep("indianred",3)))
```


```{r joint QC aftr}
ggplot(raw.data.woDbl.qcFilt@meta.data, 
       aes(x=nCount_RNA, y=nFeature_RNA , 
           color=percent.mito)) +
  geom_point() +
  scale_x_log10(labels = scales::comma) +
  scale_y_log10(labels = scales::comma) +
  theme_bw() +
  ggtitle("Features detected vs Library Size")  +
  facet_wrap(~orig.ident) 
```

### Discard non-expressed and ribosomal genes

Finally, those genes that are expressed in less than `r min_cells` (considering all samples together) will be discarded. The number of cells expressing a particular gene, per sample, is computed:

```{r compute n cells expressing a particular gene}
# Are there any genes with zero expression in all cells?
samples <- c("WT_s1", "WT_s2", "WT_s3", "g7_s1", "g7_s2", "g7_s3")
slot <- paste0("counts.", samples)

n_cells <- sapply(slot, function(case) Matrix::rowSums(raw.data.woDbl.qcFilt@assays$RNA[case] > 0))

knitr::kable(head(n_cells, n=10), caption="Example of number of cells expressing a particular gene (in rows) per sample (columns)")
  
```

```{r discard genes}
total_cells <- rowSums(n_cells) 
kept_genes <- names(total_cells[which(total_cells > min_cells)])
raw.data.woDbl.qcFilt.genes.Filt <- subset(raw.data.woDbl.qcFilt, features = kept_genes)
```

As a reminder, a total of `r nrow(n_cells)` genes were initially included in the expression matrix (for all samples). Out of those, `r length(kept_genes)` are kept for downstream analysis, after discarding the non-expressed ones (or just residual).

Finally, ribosomal genes are removed from this dataset. The number of present ribosomal genes are `r length(grep('^Rp[sl]', rownames(raw.data.woDbl.qcFilt.genes.Filt)))` genes.

```{r store object}
non.ribo.genes <- rownames(raw.data.woDbl.qcFilt.genes.Filt)[grep('^Rp[sl]', rownames(raw.data.woDbl.qcFilt.genes.Filt),invert = TRUE)]
raw.data.woDbl.qcFilt.genes.Filt.woRibo <- subset(raw.data.woDbl.qcFilt.genes.Filt, 
                                                  features = non.ribo.genes)
saveRDS(object = raw.data.woDbl.qcFilt.genes.Filt.woRibo, file = "RDS/scRNAseq_EBs_WT_7g_After_QC_Filt_woRibo.RDS")
```

The **final dataset includes a total of `r nrow(raw.data.woDbl.qcFilt.genes.Filt.woRibo)` features (genes) over `r ncol(raw.data.woDbl.qcFilt.genes.Filt.woRibo)` cells distributed in 6 samples from two conditions.**

# Session Information

```{r}
sessionInfo()
```

# References
