---
title: "scRNAseq data analysis CRISPR Screening BM: Filtering low quality cells and genes - QC and doublet removal"
subtitle: 'Researcher: Luís Galan'
author: "Author: Eric Canton - María Maqueda"
date: "13th August 2024"
bibliography: "/home/user/Documents/Files/Projects/BM_scRNAseq/aux_files/references_scRNAseq_Analysis.bib"
output:
  html_document:
    df_print: paged
    css: /home/user/Documents/Files/Projects/BM_scRNAseq/aux_files/style.css
    toc: yes
    toc_depth: '3'
    toc_float: yes
    theme: cerulean
    highlight: haddock
  pdf_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
---

```{r packages, include=FALSE}
# Included any additional libraries required i.e. openxlsx
libraries <- c("knitr", "ggplot2", "data.table", "dplyr", "stringr", "ggpubr", "Seurat") 
check.libraries <- is.element(libraries, installed.packages()[, 1])==FALSE
libraries.to.install <- libraries[check.libraries]
if (length(libraries.to.install!=0)) {
  require(libraries.to.install)
}

success <- sapply(libraries,require, quietly = FALSE,  character.only = TRUE)
if(length(success) != length(libraries)) {stop("A package failed to return a success in require() function.")}
```

```{r general, include=FALSE}
# General settings
knitr::opts_chunk$set(message=FALSE, warning=FALSE, echo=FALSE, fig.pos="H")
options(scipen=999)

# Define colors for conditions: WT and 7g
cond.colors <- c("CD45_4_5" = "gray",
                 "CD45_BFP_4_5" = "gold3")
```

# Scope

This document aims to filter out poor quality cells and genes. This is an standard first QC step for the scRNAseq data pre-processing analysis. Commonly used QC metrics include the exploration of number of unique genes detected in each cell or the percentage of reads that map to the mitochondrial genome.

# Primary data

A total of **two samples** (Bone Marrow) were library-prepared and sequenced by CRG Sequencing platform at PRBB. Libraries were prepared using **10x Single Cell 3'v3 kit**. Corresponding biological samples were delivered to the sequencing platform in December'24 and sequencing data was received in February'24. Each sample corresponds to a different condition:

-   CD45+ cells at 4.5 months (**CD45_4_5**)
-   CD45+ cells transplanted from KDR+ progenitors at 4.5 months (**CD45_BFP_4_5**) activated from 7* specific genes identified by CRISPRa in an earlier project phase.

*Genes related to HSC development fate.

# Methods

Following criteria is applied:

-   Cells with mitochondrial content higher than **12%** are discarded.
-   Cells with less than 500 counts are discarded.
-   Cells with more than 10k genes or less than 300 genes are discarded.

Ribosomal genes are removed from the dataset. Additionally, those genes expressed in 5 or less cells (among all samples) are also discarded.

QC is performed by means of Seurat R package [@SeuratV5] (v5.0.0).

REMARKS: (a) Applied criteria is quite permissive since, in general, it is expected to conduct a second round of QC (b) Cells classified as doublets/multiplets per sample, identified in a previous analysis, will be also inspected.

```{r}
# QC criteria
min_nCount <- 500
min_n_genes <- 300
max_n_genes <- 10000
max_mito <- 12
min_cells <- 5
```

# Results

All scRNAseq data samples were merged into a Seurat object

```{r import Seurat object}
raw.data <- readRDS(file = "/home/user/Documents/Files/Projects/BM_scRNAseq/RDS/scRNAseq_CRISPR_Screening_BM_RAW_original_data.RDS")

raw.data.list <- SplitObject(raw.data, split.by = "orig.ident")

raw.data <- merge(raw.data.list$CD45_4_5, raw.data.list$CD45_BFP_4_5)

raw.data
```

The complete dataset, without any filtering applied, includes a total of **`r nrow(raw.data)` features (genes)** over **`r ncol(raw.data)` samples (cells)** distributed in the three samples.

```{r cells distribution}
qc.summary <- as.data.frame(table(raw.data$Sample_ID))
colnames(qc.summary) <- c("Sample.ID", "Initial.Cells")
```

## Doublets removal

Doublets will not be discarded at this point. Their assessment will be conducted in next analysis phase.

```{r doublets removal}
raw.data.woDbl <- subset(x =  raw.data, subset = scDblFinder.class == "singlet")
qc.summary$Cells.woDBL <- table(raw.data.woDbl$Sample_ID)

knitr::kable(qc.summary, caption="Number of cells per sample: initially and after doublets removal")
```

## QC metrics

```{r include QC proportions in metadata}

# MITOCHONDRIAL genes
raw.data.woDbl <- Seurat::PercentageFeatureSet(raw.data.woDbl, 
                                    pattern = "^mt-", 
                                    col.name = "percent.mito")

# RIBOSOMAL genes
raw.data.woDbl <- Seurat::PercentageFeatureSet(raw.data.woDbl, 
                                    pattern = "^Rp[sl]",
                                    col.name = "percent.ribo")

```

### Exploration

Prior to apply any QC filtering, first a general exploration is conducted to assess the overall cells quality. This includes, per barcode (cell):

-   Number of genes
-   Number of counts
-   Mitochondrial content
-   Ribosomal content

```{r QC exploration 1}
Seurat::VlnPlot(raw.data.woDbl, layer = "counts",
        features = c("nFeature_RNA", "nCount_RNA"), 
        alpha=0.1, 
        ncol = 2, 
        cols=cond.colors)
```

```{r QC exploration 2}
Seurat::VlnPlot(raw.data.woDbl, layer = "counts",
        features = c("percent.mito", "percent.ribo"), 
        alpha=0.1, 
        ncol = 2, 
        cols=cond.colors)
```

Alternatively, a joint plot between metrics is also of interest to check relationships among variables.

```{r joint QC}
ggplot(raw.data.woDbl@meta.data, 
       aes(x=nCount_RNA, y=nFeature_RNA , 
           color=percent.mito)) +
  geom_point() +
  scale_x_log10(labels = scales::comma) +
  scale_y_log10(labels = scales::comma) +
  theme_bw() +
  ggtitle("Features detected vs Library Size")  +
  facet_wrap(~orig.ident) 
```

All samples behave roughly equally independently of their condition.

### Subset low quality cells

Cells are subset based on above mentioned QC criteria. The final number of cells is shown in the following table.

```{r subset low quality cells}
is_low_quality <- 
  raw.data.woDbl$nCount_RNA < min_nCount |
  raw.data.woDbl$nFeature_RNA < min_n_genes |
  raw.data.woDbl$nFeature_RNA > max_n_genes |
  raw.data.woDbl$percent.mito > max_mito 

#table(is_low_quality)
raw.data.woDbl.qcFilt <- subset(raw.data.woDbl, cells = colnames(raw.data.woDbl)[!is_low_quality])

qc.summary$Cells.FINAL <- table(raw.data.woDbl.qcFilt$Sample_ID)
knitr::kable(qc.summary, caption="Number of cells per sample: initially, after doublets removal and after final QC criteria")
```

After filtering, QC plots are again visualized:

```{r QC exploration 1 after}
Seurat::VlnPlot(raw.data.woDbl.qcFilt, layer = "counts",
        features = c("nFeature_RNA", "nCount_RNA"), 
        alpha=0.1, 
        ncol = 2, 
        cols=cond.colors)
```

```{r QC exploration 2 after}
Seurat::VlnPlot(raw.data.woDbl.qcFilt, layer = "counts",
        features = c("percent.mito", "percent.ribo"), 
        alpha=0.1, 
        ncol = 2, 
        cols=cond.colors)
```

```{r joint QC aftr}
ggplot(raw.data.woDbl.qcFilt@meta.data, 
       aes(x=nCount_RNA, y=nFeature_RNA , 
           color=percent.mito)) +
  geom_point() +
  scale_x_log10(labels = scales::comma) +
  scale_y_log10(labels = scales::comma) +
  theme_bw() +
  ggtitle("Features detected vs Library Size")  +
  facet_wrap(~orig.ident) 
```

### Discard non-expressed and ribosomal genes

Finally, those genes that are expressed in less than `r min_cells` (considering all samples together) will be discarded. The number of cells expressing a particular gene, per sample, is computed:

```{r compute n cells expressing a particular gene}
# Are there any genes with zero expression in all cells?
slot <- paste0("counts.", names(cond.colors))

n_cells <- sapply(slot, function(case) Matrix::rowSums(raw.data.woDbl.qcFilt@assays$RNA[case] > 0))

knitr::kable(head(n_cells, n=10), caption="Example of number of cells expressing a particular gene (in rows) per sample (columns)")
  
```

```{r discard genes}
total_cells <- rowSums(n_cells) 
kept_genes <- names(total_cells[which(total_cells > min_cells)])
raw.data.woDbl.qcFilt.genes.Filt <- subset(raw.data.woDbl.qcFilt, features = kept_genes)
```

As a reminder, a total of `r nrow(n_cells)` genes were initially included in the expression matrix (for all samples). Out of those, `r length(kept_genes)` are kept for downstream analysis, after discarding the non-expressed ones (or just residual).

Finally, **ribosomal genes are removed from this dataset**. The number of present ribosomal genes are `r length(grep('^Rp[sl]', rownames(raw.data.woDbl.qcFilt.genes.Filt)))` genes.

```{r store object 2}
non.ribo.genes <- rownames(raw.data.woDbl.qcFilt.genes.Filt)[grep('^Rp[sl]', rownames(raw.data.woDbl.qcFilt.genes.Filt),invert = TRUE)]
raw.data.woDbl.qcFilt.genes.Filt.woRibo <- subset(raw.data.woDbl.qcFilt.genes.Filt, 
                                                  features = non.ribo.genes)

saveRDS(object = raw.data.woDbl.qcFilt.genes.Filt.woRibo, file = "/home/user/Documents/Files/Projects/BM_scRNAseq/RDS/scRNAseq_CRISPR_Screening_BM_After_QC_and_dbl_removal_QC2_Only2Samples.RDS")
```

The **final dataset includes a total of `r nrow(raw.data.woDbl.qcFilt.genes.Filt.woRibo)` features (genes) over `r ncol(raw.data.woDbl.qcFilt.genes.Filt.woRibo)` cells distributed in two conditions. Doublet cells have been removed.**

# Session Information

```{r}
sessionInfo()
```

# References
